agent/client/cache/ring.go

```diff
@@ -0,0 +1,219 @@
+// Copyright 2019 Percona LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package cache encapsulates client cache logic.
+package cache
+
+import (
+	"context"
+	"strings"
+
+	"github.com/cloudflare/buffer"
+	"github.com/sirupsen/logrus"
+	"google.golang.org/protobuf/encoding/prototext"
+	"google.golang.org/protobuf/proto"
+
+	"github.com/percona/pmm/agent/config"
+	"github.com/percona/pmm/agent/models"
+	"github.com/percona/pmm/api/agentpb"
+)
+
+type Cacher interface {
+	Send(models.AgentResponse)
+	SendAndWaitResponse(agentpb.AgentRequestPayload) (agentpb.ServerResponsePayload, error)
+	SendToStream(send func(*agentpb.AgentMessage) error)
+}
+
+const drainBytesAmount = 1024 * 1024
+
+// Ring represent cache implementation based on ring buffer.
+type Ring struct {
+	ctx context.Context //nolint:containedctx
+	l   *logrus.Entry
+	// primary buffer to store valuable agent messages
+	primary *buffer.Buffer
+	// secondary buffer to store mostly QAN agent messages
+	secondary *buffer.Buffer
```

Comment:
It's like one of the solutions that came to mind. I don't insist on this specific implementation... but it could look like this for example. In order to be able to make a more appropriate decision based on this. It's just that we have a need to have important messages not overwritten by unimportant messages.